### PR TITLE
Next: Custom and overwritable CT GraphQL queries/variables

### LINKED ISSUE #4763: Make GraphQL queries overrideable
## What is the motivation for adding / enhancing this feature?

Right now queries in Commercetools integration are setted in stone which is notably limiting benefits of using GraphQL in the project. Users should be able to modify the queries per composable.

## Proposed API

In the API Client (`packages/commercetools/api-client`) the custom GraphQL query (and query only) should be the last parameter of a function. If it's not provided we're using the default one (so the behavior remains unchanged). 

```js
getProduct({ params}, customQuery)
```

On the composables level we should make this behavior availalable through `search` and `load` functions. Also we won't implement that in all composables since some of them has to be refactored. Below is a list of new APIs in composables

- `useProduct`, `useCategory`
```js
search(params, customQuery)
```
- `useCart`, `useUser`, `useUserOrders`, `useUserAddresses`
```js
load(customQuery)
```
In that case `customQuery` has to be added to commercetools integration typings and to `factoryParams` in `core` using functions overloading (not all integrations has GraphQL so its not a typing that will appear on every integration). Feel free to suggest better approach (like inheritance)

## What are the acceptance criteria 

- [ ] All API Client queries are overrideable
- [ ] Listed Composables are allowing to override their API Client method queries
- [ ] TypeScript Interfaces are extended
- [ ] A document about extending queries is added to the composables

## Version of Vue Storefront

- [ ] Vue Storefront
- [x] Vue Storefront Next

## Can you complete this feature request by yourself?

- [ ] YES
- [x] NO

### PR DESCRIPTION
**Pull request by lukasborawski**
### Related Issues

closes #4763 

### Short Description and Why It's Useful

You can now pass custom GraphQL query and variables to get some search data.

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [x] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: packages/commercetools/api-client/src/api/getProduct/index.ts
Comment (filrak) [inline packages/commercetools/api-client/src/api/getProduct/index.ts:11]:
```suggestion
const getProduct = async (params, customQuery = async (query = defaultQuery, variables = {}) => {
```
according to the interface in the issue

File: packages/commercetools/api-client/src/api/getProduct/index.ts
Comment (filrak) [inline packages/commercetools/api-client/src/api/getProduct/index.ts:2]: will conflict with https://github.com/DivanteLtd/vue-storefront/pull/4791 - please wait until its merged and correct accordingly

File: packages/commercetools/api-client/src/api/getProduct/index.ts
Comment (filrak) [inline packages/commercetools/api-client/src/api/getProduct/index.ts:15]: if the variables are now configurable you have to make sure that if some of them will be removed the code will still work. Try to be more [defensive](https://enterprisecraftsmanship.com/posts/defensive-programming/) here ;)

File: packages/core/core/src/types.ts
Comment (filrak) [inline packages/core/core/src/types.ts:18]: this will display `customQuery` for each integration while GQL is not a part of each

File: packages/commercetools/composables/src/useProduct/index.ts
Comment (filrak) [inline packages/commercetools/composables/src/useProduct/index.ts:32]: thI think we can just type the interface of `customQuery` more loosely 9without `apolloQueryResult` and remove it from the generic interface (so have it dictated by the core, not integration developers)

File: packages/commercetools/api-client/src/api/getProduct/index.ts
Comment (lukasborawski) [reply packages/commercetools/api-client/src/api/getProduct/index.ts:2]: ok, will wait for it

File: packages/commercetools/composables/src/useProduct/index.ts
Comment (lukasborawski) [reply packages/commercetools/composables/src/useProduct/index.ts]: Done.

File: packages/core/core/src/types.ts
Comment (lukasborawski) [reply packages/core/core/src/types.ts]: Done.

File: packages/commercetools/api-client/src/api/getProduct/index.ts
Comment (lukasborawski) [reply packages/commercetools/api-client/src/api/getProduct/index.ts]: Done.

File: packages/commercetools/api-client/src/api/getProduct/index.ts
Comment (lukasborawski) [reply packages/commercetools/api-client/src/api/getProduct/index.ts]: @filrak Actually this is more like resolving the final query then custom query for all cases. But we can leave it now, especially that we have multiple implementations now. 😃

File: packages/commercetools/composables/src/useCategory/index.ts
Comment (filrak) [inline packages/commercetools/composables/src/useCategory/index.ts:7]: `customQuery` should be optional, same for product

File: packages/commercetools/api-client/src/helpers/search/index.ts
Comment (lukasborawski) [reply packages/commercetools/api-client/src/helpers/search/index.ts]: Done.

File: packages/commercetools/composables/src/useCategory/index.ts
Comment (lukasborawski) [reply packages/commercetools/composables/src/useCategory/index.ts:7]: Done.

File: packages/core/core/src/types.ts
Comment (lukasborawski) [reply packages/core/core/src/types.ts]: It's already there.

File: packages/core/docs/commercetools/composables.md
Comment (filrak) [inline packages/core/docs/commercetools/composables.md:43]:
```suggestion
const customQuery = (query, variables) => {
  // import a custom Graphql QUERY
  const newQuery = require('./queries/customQuery.gql')
  // override "locale' variable and leave other ones untouched
  const newVariables = {  ..variabes, locale: 'en' }
  
  return {
    query: newQuery,
    variables: newVariables
  }
}
search({ id: '12345' }, customQuery)
```

File: packages/core/docs/core/composables.md
Comment (filrak) [inline packages/core/docs/core/composables.md:47]:
```suggestion
There's an option to use your own GraphQL query and variables while searching.   
```

File: packages/core/docs/core/composables.md
Comment (lukasborawski) [reply packages/core/docs/core/composables.md]: Done.

File: packages/core/docs/commercetools/composables.md
Comment (lukasborawski) [reply packages/core/docs/commercetools/composables.md]: Done.

File: packages/core/docs/core/composables.md
Comment (lukasborawski) [reply packages/core/docs/core/composables.md]: Done.

**Review by andrzejewsky — CHANGES_REQUESTED** [review]
Looks good, one improvement:

```ts
interface CustomQueryParameters<V> {
  query?: any;
  variables: V;
}

interface CustomQuery<V> {
 query: any;
 variables: V;
}

type CustomQueryFn = <T = any>(params: CustomQueryParameters<T>) => CustomQuery<T>
```

function signature

```ts
const getCategory = async <T = any>(params, customQuery: CustomQueryFn<T>) => {
 const { query, variables } = customQuery(defaultQuery); // selects the default "query" if query has not driven
 ...
}
```

**Comment by andrzejewsky** [comment]
next step: https://github.com/DivanteLtd/vue-storefront/issues/4844

**Review by andrzejewsky — APPROVED** [review]

**Pull request merged**
